Sources/Datadog/Core/System/MobileDevice.swift

```diff
@@ -74,28 +78,51 @@ internal class MobileDevice {
             }
         )
     }
+
+    // MARK: - Singleton
+    private static var _instance: MobileDevice?
+
     /// Returns current mobile device  if `UIDevice` is available on this platform.
     /// On other platforms returns `nil`.
     static var current: MobileDevice {
-        #if !targetEnvironment(simulator)
-        // Real device
-        return MobileDevice(
-            uiDevice: UIDevice.current,
-            processInfo: ProcessInfo.processInfo,
-            notificationCenter: .default
-        )
-        #else
-        // iOS Simulator - battery monitoring doesn't work on Simulator, so return "always OK" value
-        return MobileDevice(
-            model: UIDevice.current.model,
-            osName: UIDevice.current.systemName,
-            osVersion: UIDevice.current.systemVersion,
-            enableBatteryStatusMonitoring: {},
-            resetBatteryStatusMonitoring: {},
-            currentBatteryStatus: { BatteryStatus(state: .full, level: 1, isLowPowerModeEnabled: false) }
-        )
-        #endif
+        get {
+            if let instance = _instance {
+                return instance
+            }
+
+            #if !targetEnvironment(simulator)
+            // Real device
+            _instance = MobileDevice(
+                uiDevice: UIDevice.current,
+                processInfo: ProcessInfo.processInfo,
+                notificationCenter: .default
+            )
+            #else
+            // iOS Simulator - battery monitoring doesn't work on Simulator, so return "always OK" value
+            _instance = MobileDevice(
+                model: UIDevice.current.model,
+                osName: UIDevice.current.systemName,
+                osVersion: UIDevice.current.systemVersion,
+                processInfo: ProcessInfo.processInfo,
+                enableBatteryStatusMonitoring: {},
+                resetBatteryStatusMonitoring: {},
+                currentBatteryStatus: { BatteryStatus(state: .full, level: 1, isLowPowerModeEnabled: false) }
+            )
+            #endif
+
+            // swiftlint:disable:next force_unwrapping
+            return _instance!
+        }
+        set(newInstance) {
+            _instance = newInstance
+        }
+    }
+
+    #if DD_SDK_COMPILED_FOR_TESTING
+    static func clearForTesting() {
+        _instance = nil
     }
+    #endif
```

Comment:
I think we can simplify this and avoid having to use a second singleton by creating another `convenience init`, like:
```swift
    convenience init() {
        #if !targetEnvironment(simulator)
        // Real device
        self.init(
            uiDevice: UIDevice.current,
            processInfo: ProcessInfo.processInfo,
            notificationCenter: .default
        )
        #else
        // iOS Simulator - battery monitoring doesn't work on Simulator, so return "always OK" value
        self.init(
            model: UIDevice.current.model,
            osName: UIDevice.current.systemName,
            osVersion: UIDevice.current.systemVersion,
            processInfo: ProcessInfo.processInfo,
            enableBatteryStatusMonitoring: {},
            resetBatteryStatusMonitoring: {},
            currentBatteryStatus: { BatteryStatus(state: .full, level: 1, isLowPowerModeEnabled: false) }
        )
        #endif
    }

    /// Returns current mobile device.
    static var current = MobileDevice()

    #if DD_SDK_COMPILED_FOR_TESTING
    static func clearForTesting() {
        current = MobileDevice()
    }
    #endif
```

Also the `clearForTesting` might not be even necessary if we do:
```swift
func testSomething() {
    MobileDevice.current = .mock()
    defer { MobileDevice.current = .init() }

    ...
}
```

Reply:
That's a good call. I was originally thinking ok making `set` only visible to testing, but that wasn't really possible so your solution ends up being a lot cleaner given what's possible.